File: Python/studentGrade.py
def input_student_scores(all_students, no_of_students, no_of_subjects):
	all_students = []
	student_counter = 1

	for _ in range(no_of_students):
		each_student = []
		total = 0
		print(f"Entering score for student {student_counter}")
		for subject_score in range(no_of_subjects):
			while True:
				try:
					score = float(input(f"Enter score for subject {subject_score + 1}: "))
					if 0 <= score <= 100:
						total += score
						each_student.insert(subject_score, score)
						break
					else:
						print("Invalid score.\nScore must be between 0 and 100!")
				except:
					print("You may have entered an invalid character. Retry")

		each_student.append(total)
		each_student.append(total / no_of_subjects)
		each_student.append(0)  # position
		all_students.append(each_student)
		print("Saving >>>>>>>>>>>>>>>>\nSaved successfully")
		student_counter += 1

	return all_students


def position_of_students(all_students):
	average_scores_of_each_student = []

	for each_student in all_students:
		average_scores_of_each_student.append(each_student[-2])

	for eachIndex in range(len(average_scores_of_each_student) - 1):
		for spanIndex in range(eachIndex + 1, len(average_scores_of_each_student)):
			if average_scores_of_each_student[eachIndex] < average_scores_of_each_student[spanIndex]:
				average_scores_of_each_student[eachIndex], average_scores_of_each_student[spanIndex] = (
				average_scores_of_each_student[spanIndex],
				average_scores_of_each_student[eachIndex],)

	student_has_position = []

	for each_student in all_students:
		for highest_to_lowest in range(len(average_scores_of_each_student)):
			if each_student[-2] == average_scores_of_each_student[highest_to_lowest] and each_student not in student_has_position:
				each_student[-1] = highest_to_lowest + 1
				student_has_position.append(each_student)

	return all_students


def display_record_sheet(all_students, no_of_subjects):
	print("===============================================")
	header = ["STUDENT"]
	for index in range(1, no_of_subjects + 1):
		header.append(f"SUB{index}")
	header += ["TOTAL", "AVE", "POS"]

	for each_header in header:
		print(each_header, end="\t")
		if each_header == "STUDENT":
			print("\t", end="")

	print("\n" + "=" * 90)

	student_counter = 1
	for each_student_and_scores in all_students:
		print(f"Student {student_counter}", end="\t")
		for index, value in enumerate(each_student_and_scores):
			if index == len(each_student_and_scores) - 1:
				print(int(value), end="")
			elif index == len(each_student_and_scores) - 2:
				print(f"{value:.2f}", end="\t")
			else:
				print(value, end="\t")
		print()
		student_counter += 1

	print("=" * 90)
	print("=" * 90)


def subject_summary(all_students, no_of_subjects, no_of_students):

	for each_subject_index in range(no_of_subjects):
		
		highest = float('-inf')
		lowest = float('inf')
		highest_scoring_student = 0
		lowest_scoring_student = 0
		subject_total = 0
		pass_count = 0
		fail_count = 0
		for index, each_student_score_in_subject in enumerate(all_students):
			score = each_student_score_in_subject[each_subject_index]
			if score > highest:
				highest = score
				highest_scoring_student = index + 1
			if score < lowest:
				lowest = score
				lowest_scoring_student = index + 1
			if score >= 50:
				pass_count += 1
			else:
  				fail_count += 1
			subject_total += score

		print(f"Subject {each_subject_index + 1}")
		print(f"Highest scoring student is: Student {highest_scoring_student} scoring {highest:.2f}.")
		print(f"Lowest scoring student is: Student {lowest_scoring_student} scoring {lowest:.2f}.")
		print(f"Total score is: {subject_total:.2f}.")
		print(f"Average score is: {subject_total / no_of_students:.2f}.")
		print(f"Number of passes: {pass_count}.")
		print(f"Number of failures: {fail_count}.")


def class_summary(all_students, no_of_subjects, no_of_students):
	compare_subject_failures = 0
	compare_subject_passes = 0
	overall_highest_subject_score = float('-inf')
	overall_lowest_subject_score = float('inf')
	all_students_index = 0
	highest_score_in_all_index = 0
	subject_with_highest_score = 0
	lowest_score_in_all = 0
	subject_with_lowest_score = 0
	sum_student_total = 0
	highest_students_total = float('-inf')
	lowest_students_total = float('inf')
	best_graduating_student = 0
	worst_graduating_student = 0
	hardest_subject_index = 0
	easiest_subject_index = 0
	class_total_score = 0
	count_no_of_passes = 0
	count_no_of_failures = 0

	for each_subject_index in range(no_of_subjects):
		for each_student_score_in_subject in all_students:
			if each_student_score_in_subject[each_subject_index] > overall_highest_subject_score:
				overall_highest_subject_score = each_student_score_in_subject[each_subject_index]
				highest_score_in_all_index = all_students_index + 1
				subject_with_highest_score = each_subject_index + 1

			if each_student_score_in_subject[each_subject_index] < overall_lowest_subject_score:
				overall_lowest_subject_score = each_student_score_in_subject[each_subject_index]
				lowest_score_in_all = all_students_index + 1
				subject_with_lowest_score = each_subject_index + 1

			if each_student_score_in_subject[each_subject_index] >= 50:
				count_no_of_passes += 1
			else:
				count_no_of_failures += 1

			if each_subject_index == 0:
				for add_scores_of_student in range(no_of_subjects):
					sum_student_total += each_student_score_in_subject[add_scores_of_student]
				class_total_score += sum_student_total

				if sum_student_total > highest_students_total:
					highest_students_total = sum_student_total
					best_graduating_student = all_students_index + 1
				if sum_student_total < lowest_students_total:
					lowest_students_total = sum_student_total
					worst_graduating_student = all_students_index + 1

			all_students_index += 1
			sum_student_total = 0

		if each_subject_index == 0:
			compare_subject_failures = count_no_of_failures
			compare_subject_passes = compare_subject_passes
			hardest_subject_index = each_subject_index + 1
			easiest_subject_index = each_subject_index + 1
		else:
			if count_no_of_passes > compare_subject_passes:
				compare_subject_passes = count_no_of_passes
				easiest_subject_index = each_subject_index + 1
			elif count_no_of_failures > compare_subject_failures:
				compare_subject_failures = count_no_of_failures
				hardest_subject_index = each_subject_index + 1

		count_no_of_failures = 0
		count_no_of_passes = 0
		all_students_index = 0

	if compare_subject_passes > 0 and compare_subject_failures > 0:
		if compare_subject_failures != compare_subject_passes:
			print(f"\n\nThe hardest subject is subject {hardest_subject_index} with {compare_subject_failures} failures.")
			print(f"The easiest subject is subject {easiest_subject_index} with {compare_subject_passes} passes.")
			print(f"The overall highest score is scored by student {highest_score_in_all_index} in subject {subject_with_highest_score} scoring {overall_highest_subject_score:.2f}.")
			print(f"The overall lowest score is scored by student {lowest_score_in_all} in subject {subject_with_lowest_score} scoring {overall_lowest_subject_score:.2f}.")
		else:
			print("There were equal number of passes and failures in all subjects.")
			print(f"The overall highest score is scored by student {highest_score_in_all_index} in subject {subject_with_highest_score} scoring {overall_highest_subject_score:.2f}.")
			print(f"The overall lowest score is scored by student {lowest_score_in_all} in subject {subject_with_lowest_score} scoring {overall_lowest_subject_score:.2f}.")

		print(f"\n\n\nCLASS SUMMARY\n\n")
		print(f"Best graduating student is: student {best_graduating_student} scoring {highest_students_total:.2f}.")
		print(f"Worst graduating student is: student {worst_graduating_student} scoring {lowest_students_total:.2f}.\n\n\n")
	elif compare_subject_passes > 0 and compare_subject_failures == 0:
		print("No failures")
	elif compare_subject_passes == 0 and compare_subject_failures > 0:
		print("All students failed all subjects")

	print(f"Class total score is {class_total_score:.2f}.")
	print(f"Class average score is {class_total_score / no_of_students:.2f}.")



run_app = True

while run_app:
	try:
		no_of_students = input("How many students do you have? ")
		if not no_of_students.isdigit():
			raise ValueError("You may have entered an invalid character.")
		convert_stu = int(no_of_students)
		if convert_stu < 1:
			raise ValueError("This field requires a positive integer.")

		no_of_subjects = input("How many subjects do they offer? ")
		if not no_of_subjects.isdigit():
			raise ValueError("You may have entered an invalid character.")
		convert_sub = int(no_of_subjects)
		if convert_sub < 1:
			raise ValueError("This field requires a positive integer. Retry.")

		all_students = []
		all_students = input_student_scores(all_students, convert_stu, convert_sub)
		all_students = position_of_students(all_students)
		display_record_sheet(all_students, convert_sub)
		subject_summary(all_students, convert_sub, convert_stu)
		class_summary(all_students, convert_sub, convert_stu)
		run_app = False

	except ValueError as invalid_input:
		print(invalid_input)

File: JavaScript/studentGrade.js
const prompt = require('prompt-sync')();

function inputStudentScores(allStudents, noOfStudents, noOfSubjects) {
	let studentCounter = 1;
	allStudents = [];

	for(let index = 0; index < noOfStudents; index++){
		let eachStudent = [];
		let total = 0;

		for(let subjectScore = 0; subjectScore < noOfSubjects; subjectScore++) {
			while(true){
				try{
					console.log("Entering score for student " + studentCounter);
					let score = parseFloat(prompt("Enter score for subject " + (subjectScore + 1) + ": "));
					if(isNaN(score))throw new Error("You may have entered an invalid character. Retry.");
					if (score >= 0 && score <= 100){
						total += score;
						eachStudent[subjectScore] = score;
						break;
					} else {
						console.log("Invalid score.\nScore must be between 0 and 100!");
					}
				} catch(invalid){
					console.log(invalid.message);
				}
			}
		}

		eachStudent[noOfSubjects] = total;
		eachStudent[noOfSubjects + 1] = total / noOfSubjects;
		eachStudent[noOfSubjects + 2] = 0; // placeholder for position
		allStudents.push(eachStudent);
		console.log("Saving >>>>>>>>>>>>>>>>\nSaved successfully");
		studentCounter++;
	}

	return allStudents;
}

function postionOfStudents(allStudents) {
	let averageScoresOfEachStudent = [];

	for(let index = 0; index < allStudents.length; index++){
		let eachStudent = allStudents[index];
		averageScoresOfEachStudent.push(eachStudent[eachStudent.length - 2]); //adds each studen average score to a list.
	}

	for(let index = 0; index < averageScoresOfEachStudent.length - 1; index++){
		for (let spanIndex = index + 1; spanIndex < averageScoresOfEachStudent.length; spanIndex++) {
			if(averageScoresOfEachStudent[index] < averageScoresOfEachStudent[spanIndex]) {
				let tempHold = averageScoresOfEachStudent[index];
				averageScoresOfEachStudent[index] = averageScoresOfEachStudent[spanIndex];
				averageScoresOfEachStudent[spanIndex] = tempHold;
			}
		}
	}

	let studentHasPosition = [];

	for (let index = 0; index < allStudents.length; index++) {
		let eachStudent = allStudents[index];

		for (let highestToLowest = 0; highestToLowest < averageScoresOfEachStudent.length; highestToLowest++) {
			if (eachStudent[eachStudent.length - 2] === averageScoresOfEachStudent[highestToLowest] &&
				!studentHasPosition.includes(eachStudent)) {
				eachStudent[eachStudent.length - 1] = highestToLowest + 1;
				studentHasPosition.push(eachStudent);
			}
		}
	}

	return allStudents;
}

function displayRecordSheet(allStudents, noOfSubjects){
	console.log("===============================================");

	let header = [];
	header[0] = "STUDENT";
	for (let index = 1; index <= noOfSubjects; index++) {
		header[index] = "SUB" + index;
	}
	header[header.length] = "TOTAL";
	header[header.length] = "AVE";
	header[header.length] = "POS";

	for(let eachHeader of header){
		process.stdout.write(eachHeader + "\t");
		if(header[0] === eachHeader) process.stdout.write("\t");
	}
	console.log("\n==============================================================================================");

	let studentCounter = 1;
	for (let indexOfEach = 0; indexOfEach < allStudents.length; indexOfEach++) {
		let eachStudentAndScores = allStudents[indexOfEach];
		process.stdout.write("Student " + studentCounter + "\t");

		for (let index = 0; index < eachStudentAndScores.length; index++) {
			let printEach = eachStudentAndScores[index];
			if (eachStudentAndScores[eachStudentAndScores.length - 1] === printEach)
				process.stdout.write(parseInt(printEach) + "");
			else if (eachStudentAndScores[eachStudentAndScores.length - 2] === printEach)
				process.stdout.write(printEach.toFixed(2) + "\t");
			else
				process.stdout.write(printEach + "\t");
		}

		studentCounter++;
		console.log();
	}
	console.log("==========================================================================================");
	console.log("==========================================================================================");
}

function subjectSummary(allStudents, noOfSubjects, noOfStudents) {

	for (let eachSubjectIndex = 0; eachSubjectIndex < noOfSubjects; eachSubjectIndex++){

		let highest = Number.MIN_SAFE_INTEGER;
		let lowest = Number.MAX_SAFE_INTEGER;
		let indexOfHighest = 0;
		let highestScoringStudent = 0;
		let indexOfLowest = 0;
		let lowestScoringStudent = 0;
		let sumSubjectTotal = 0;
		let countNoOfPasses = 0;
		let countNoOfFailures = 0;

		for (let indexOfEach = 0; indexOfEach < allStudents.length; indexOfEach++){
			let eachStudentScoreInSubject = allStudents[indexOfEach];

			if (eachStudentScoreInSubject[eachSubjectIndex] > highest){
				highest = eachStudentScoreInSubject[eachSubjectIndex];
				highestScoringStudent = indexOfHighest + 1;

			} if(eachStudentScoreInSubject[eachSubjectIndex] < lowest){
				lowest = eachStudentScoreInSubject[eachSubjectIndex];
				lowestScoringStudent = indexOfLowest + 1;
			}

			if (eachStudentScoreInSubject[eachSubjectIndex] >= 50) countNoOfPasses++;
			else{countNoOfFailures++;}

			sumSubjectTotal += eachStudentScoreInSubject[eachSubjectIndex];
			indexOfHighest++;
			indexOfLowest++;
		}

		console.log("Subject " + (eachSubjectIndex + 1));
		console.log("Highest scoring student is: Student " + highestScoringStudent + " scoring " + highest.toFixed(2) + ".");
		console.log("Lowest scoring student is: Student " + lowestScoringStudent + " scoring " + lowest.toFixed(2) + ".");
		console.log("Total score is: " + sumSubjectTotal.toFixed(2) + ".");
		console.log("Average score is: " + (sumSubjectTotal / noOfStudents).toFixed(2) + ".");
		console.log("Number of passes: " + countNoOfPasses + ".");
		console.log("Number of failures: " + countNoOfFailures + ".");

	}
}


function classSummary(allStudents, noOfSubjects, noOfStudents) {
	let compareSubjectFailures = 0;
	let compareSubjectPasses = 0;
	let overallHighestSubjectScore = Number.MIN_SAFE_INTEGER;
	let overallLowestestSubjectScore = Number.MAX_SAFE_INTEGER;
	let allStudentsIndex = 0;
	let highestScoreInAllIndex = 0;
	let subjectWithHighestScore = 0;
	let lowestScoreInAll = 0;
	let subjectWithLowestScore = 0;
	let sumStudentTotal = 0;
	let highestStudentsTotal = Number.MIN_SAFE_INTEGER;
	let lowestStudentsTotal = Number.MAX_SAFE_INTEGER;
	let bestGraduatingStudent = 0;
	let worstGraduatingStudent = 0;
	let hardestSubjectIndex = 0;
	let easiestSubjectIndex = 0;
	let classTotalScore = 0;
	let countNoOfPasses = 0;
	let countNoOfFailures = 0;

	for(let eachSubjectIndex = 0; eachSubjectIndex < noOfSubjects; eachSubjectIndex++){

		for(let eachStudentScoreInSubject of allStudents){

			if(eachStudentScoreInSubject[eachSubjectIndex] > overallHighestSubjectScore){
				overallHighestSubjectScore = eachStudentScoreInSubject[eachSubjectIndex];
				highestScoreInAllIndex = allStudentsIndex + 1;
				subjectWithHighestScore = eachSubjectIndex + 1;
			}

			if(eachStudentScoreInSubject[eachSubjectIndex] < overallLowestestSubjectScore) {
				overallLowestestSubjectScore = eachStudentScoreInSubject[eachSubjectIndex];
				lowestScoreInAll = allStudentsIndex + 1;
				subjectWithLowestScore = eachSubjectIndex + 1;
			}

			if (eachStudentScoreInSubject[eachSubjectIndex] >= 50) {
				countNoOfPasses++;
			} else{
				countNoOfFailures++;
			}

			if(eachSubjectIndex === 0){
				for(let addScoresOfStudent = 0; addScoresOfStudent < noOfSubjects; addScoresOfStudent++){
					sumStudentTotal += eachStudentScoreInSubject[addScoresOfStudent];
				}
				classTotalScore += sumStudentTotal;

				if(sumStudentTotal > highestStudentsTotal){
					highestStudentsTotal = sumStudentTotal;
					bestGraduatingStudent = allStudentsIndex + 1;
				}
				if(sumStudentTotal < lowestStudentsTotal){
					lowestStudentsTotal = sumStudentTotal;
					worstGraduatingStudent = allStudentsIndex + 1;
				}
			}
			allStudentsIndex++;
			sumStudentTotal = 0;
		}

		if(eachSubjectIndex === 0){
			compareSubjectFailures = countNoOfFailures;
			compareSubjectPasses = compareSubjectPasses;
			hardestSubjectIndex = eachSubjectIndex + 1;
			easiestSubjectIndex = eachSubjectIndex + 1;
		} else {
			if(countNoOfPasses > compareSubjectPasses){
				compareSubjectPasses = countNoOfPasses;
				easiestSubjectIndex = eachSubjectIndex + 1;
			} else if(countNoOfFailures > compareSubjectFailures){
				compareSubjectFailures = countNoOfFailures;
				hardestSubjectIndex = eachSubjectIndex + 1;
			}
		}

		countNoOfFailures = 0;
		countNoOfPasses = 0;
		allStudentsIndex = 0;
	}

	if(compareSubjectPasses > 0 && compareSubjectFailures > 0){
		if(compareSubjectFailures > compareSubjectPasses || compareSubjectFailures < compareSubjectPasses){
			console.log(`\n\nThe hardest subject is subject ${hardestSubjectIndex} with ${compareSubjectFailures} failures.`);
			console.log(`The easiest subject is subject ${easiestSubjectIndex} with ${compareSubjectPasses} passes.`);
			console.log(`The overall highest score is scored by student ${highestScoreInAllIndex} in subject ${subjectWithHighestScore} scoring ${overallHighestSubjectScore.toFixed(2)}.`);
			console.log(`The overall lowest score is scored by student ${lowestScoreInAll} in subject ${subjectWithLowestScore} scoring ${overallLowestestSubjectScore.toFixed(2)}.`);
		} else{
			console.log("There were equal number of passes and failures in all subjects.");
			console.log(`The overall highest score is scored by student ${highestScoreInAllIndex} in subject ${subjectWithHighestScore} scoring ${overallHighestSubjectScore.toFixed(2)}.`);
			console.log(`The overall lowest score is scored by student ${lowestScoreInAll} in subject ${subjectWithLowestScore} scoring ${overallLowestestSubjectScore.toFixed(2)}.`);
		}

		console.log(`\n\n\nCLASS SUMMARY\n\n`);
		console.log(`Best graduating student is: student ${bestGraduatingStudent} scoring ${highestStudentsTotal.toFixed(2)}.`);
		console.log(`Worst graduating student is: student ${worstGraduatingStudent} scoring ${lowestStudentsTotal.toFixed(2)}.\n\n\n`);
	} else if(compareSubjectPasses > 0 && compareSubjectFailures === 0){
		console.log("No failures");
	} else if(compareSubjectPasses === 0 && compareSubjectFailures > 0){
		console.log("All students failed all subjects");
	}

	console.log(`Class total score is ${classTotalScore.toFixed(2)}.`);
	console.log(`Class average score is ${(classTotalScore / noOfStudents).toFixed(2)}.`);
}


let noOfStudents = "";
let noOfSubjects = "";
let convertStu = 0;
let convertSub = 0;
let runApp = true;

while(runApp){
	try{
		noOfStudents = prompt("How many students do you have? ");
		convertStu = parseInt(noOfStudents);
		if(isNaN(noOfStudents))throw new Error("You may have entered an invalid character. Start over");
		if(convertStu < 1) throw new Error("This field requires a positive integer.");
     

		noOfSubjects = prompt("How many subjects do they offer? ");
		convertSub = parseInt(noOfSubjects);
		if(isNaN(convertSub))throw new Error("You may have entered an invalid character. Start over");
		if(convertSub < 1)throw new Error("This field requires a positive integer. Retry.");

		let allStudents = [];
		allStudents = inputStudentScores(allStudents, convertStu, convertSub);
		allStudents = postionOfStudents(allStudents);
		displayRecordSheet(allStudents, convertSub);
		subjectSummary(allStudents, convertSub, convertStu);
		classSummary(allStudents, convertSub, convertStu);

		runApp = false;

	} catch(invalid) {
		console.log(invalid.message);
	}
}
